### PR TITLE
WIP: Multi constructor

### PR DESCRIPTION
**Pull request by btaidm**
This adds the ability to use multi-constructors in the same way PyYAML does it. This also required a refractor to pass in a Constructor as an argument.
This allows adds in more tests to cover a wider range of the code.
This would be the first step in allow Julia Types to be de-serialized from YAML files using tags.

### CODE REVIEW

**Comment by BenJWard** [comment]
Apologies for the delay reviewing this, YAML.jl tends to get pushed down the list of BioJulia projects I monitor. The tests and coverage look fine to me so I'm fine for this to be merged. @sglyon what do you think?

**Comment by sglyon** [comment]
I think this is a good change, but I believe it would break backwards compatibility downstream. 

@albop can you comment on what this would mean for Dolo.jl and Dolang.jl?

I think it is fine to break compatibility because this is a good change, we'll just need to do it in a way that gives dependencies a chance to update to the new API.

**Comment by kescobo** [comment]
Maybe just bump to `v0.4.0` on the next release? #65

**Comment by btaidm** [comment]
@sglyon I have tried my best to prevent the old methods of setting constructors/tags from breaking, but there is a chance that I missed some thing. Line 43 of the YAML.jl src file should handle the old methods.

**Comment by albop** [comment]
Sorry I just saw it. I'll check whether it has any impact on dolo.jl but I
agree the change is good in principle.

**Comment by albop** [comment]
No problem from me: dolo.jl still works. @sglyon thanks for the heads up!

**Comment by sglyon** [comment]
ok looks great, thanks for the good work here @btaidm

**Comment by alhirzel** [comment]
Can we merge this?

**Comment by kescobo** [comment]
@btaidm Sorry for the delay (again). Any chance you could resolve this new conflict (I think we just want to keep both of the conflicting lines)? I'll get it merged and make a new release. Nothing here is breaking, right?

**Comment by alhirzel** [comment]
Any word on why the CI failed?

**Comment by kescobo** [comment]
Huh... I can't look at it on travis since the move to JuliaData - can you just push something trivial to the branch to re-trigger it? Or is there an easier way I'm missing?

**Comment by kescobo** [comment]
Oh, it's here: https://travis-ci.org/github/JuliaData/YAML.jl/builds/680099610

File: test/runtests.jl [hunk 1]
```diff
@@ -133,17 +155,91 @@ end
...
+    stringData = YAML.load(yamlString, TestConstructor())
+    @test equivalent(stringData, expected)
+    dictStringData = YAML.load(yamlString, more_constructors, multi_constructors)
+    @test equivalent(dictStringData, expected)
+
+
+    @test equivalent(first(YAML.load_all_file(joinpath(testdir, string(test, ".data")), TestConstructor())), expected)
+    @test equivalent(first(YAML.load_all_file(joinpath(testdir, string(test, ".data")), more_constructors, multi_constructors)), expected)
+
+    @test equivalent(first(YAML.load_all(yamlString, TestConstructor())), expected)
```
Comment (alhirzel) [inline test/runtests.jl:185]: Failing on Windows

**Comment by btaidm** [comment]
After getting a Windows VM setup for testing the code, I ran `]test` and all the tests ran without exception. I don't know what is up with it. I am using Windows 10, build 2004. I need somebody else to look at the code.

File: test/runtests.jl (hunk 1 above)
Comment (btaidm) [reply test/runtests.jl:185]: Code works when manual running code on Windows 10, build 2004 and Julia 1.4.2

**Comment by btaidm** [comment]
I have also ran it the Bash shell the git install provides and ran all the commands that Travis-CI did, and all the tests still ran and passed.

**Comment by btaidm** [comment]
For some reason, only the `empty_scaler` test is failing and only on the CI. The tests run just fine on an up to date Windows Box.
I believe it is something to do with the CI container, but I am not sure. I have removed the `empty_scaler` test to make sure everything else ran, which it did.
@kescobo would you be able to merge this without the CI passing. I don't like the idea of removing tests just to make a single CI image happy.

**Comment by kescobo** [comment]
@btaidm Again, I need to apologize for this languishing so long. I'm not sure how exactly I ended up as the primary maintainer for this package, but here we are :facepalm:. I really appreciate your patience.

If you'll indulge me - resolve the conflicts here (you can rebase or merge, whichever you prefer), and then we'll see what happens with CI - if only windows is failing and you confirm it's working locally, I'll merge.

**Comment by btaidm** [comment]
Sorry for taking so long on getting to this. It has been a year and I am finally starting to work on some projects of mine again.

**Comment by btaidm** [comment]
I tried not to break API, but let me know if it does for any projects that are using this.

**Comment by kescobo** [comment]
Thanks for coming back to it, no worries! I'm going to ask on slack if some others with windows machines can try it out.

**Comment by kescobo** [comment]
Alright, someone on slack checked on windows and this passed. I'm a bit worried about merging with that, but maybe the switch to github ci will help?

**Comment by btaidm** [comment]
Understandable, I really wish I knew what was happening. I will review my code to double check to make sure I didn't do anything silly.

**Comment by kescobo** [comment]
Shouldn't be up to you to debug ci issues, though if you want to try switching to github ci to see if that fixes it, that would be awesome

**Comment by btaidm** [comment]
Based on the github CI, it is something with my code. I probably messed up thing with an early rebase. I am going to do a clean slate re-do of this. That might fix things.

**Comment by kescobo** [comment]
OK, thanks for your persistence! Sorry this is such a hassle :-(

**Pull request closed without merging**